src/main/java/com/fasterxml/jackson/databind/deser/DeserializerCache.java

```diff
@@ -246,12 +252,26 @@ protected JsonDeserializer<Object> _createAndCacheValueDeserializer(Deserializat
          * limitations necessary to ensure that only completely initialized ones
          * are visible and used.
          */
-        synchronized (_incompleteDeserializers) {
-            // Ok, then: could it be that due to a race condition, deserializer can now be found?
-            JsonDeserializer<Object> deser = _findCachedDeserializer(type);
-            if (deser != null) {
-                return deser;
+        if (type == null) {
```

Comment:
we can now do the _findCachedDeserializer call twice (once before lock and once inside lock) - so I have inlined the logic instead to avoid redoing some checks in _findCachedDeserializer